Initial extraction from InfluxData

randycoulman opened the pull request:
Closes #1

This contains the code almost exactly as it has been running in production at InfluxData.  We're open to make any changes or improvements necessary to get this in good shape for a wider community.

- Generated initial app skeleton:
  - `mix new . --app config_cat`
  - Don't overwrite README or .gitignore.
  - Append the README contents that would have been generated to the existing README. They're not accurate until we publish to hex.pm, but I didn't want to lose the content.

- Extract library from InfluxData's app.  I made the following changes to what we have running:
  - Add necessary dependencies to mix.exs.
  - Replace use of `Ecto.UUID.generate/0` with `UUID.uuid4/0` to avoid depending on Ecto.
  - Configure the log level in `test_helper.exs` (we had it in application config, but that's not recommended for libraries).
  - Add `@tag capture_log: true` to noisy tests. See note below.

**NOTES:**
- I'm using Elixir 1.10.4; I tried using 1.11, but we get a warning from the `elixir_uuid` library regarding its use of the `:crypto` library. Elixir 1.11 does some extra checking of this sort.  This is an open PR to address this: https://github.com/zyro/elixir-uuid/pull/47.  We'll ultimately need to decide which Elixir versions we want to support and configure those in the Travis build matrix.

- I added `mix_test_watch` as a dev-only dependency. It's very handy to be able to run `mix test.watch --stale` while developing so that tests run automatically as you make changes.

- I attempted to add some tests around the log output, but because the logging comes from a separate process, it's harder to capture that way.  If this is something we want tests around, we can do that as separate PR.

Commit:
Generate app skeleton

- Generate with `mix new . --app config_cat`
- Don't overwrite README or .gitignore
- Append the README contents that would have been generated to the existing README

Commit:
Extract library from InfluxData's app

This commit contains the code almost exactly as it has been running in production at InfluxData.

Changes here:
- Add necessary dependencies to mix.exs
- Replace use of `Ecto.UUID.generate/0` with `UUID.uuid4/0` to avoid depending on Ecto
- Configure the log level in `test_helper.exs`
- Add `@tag capture_log: true` to noisy tests.

Co-authored-by: Iris Scholten <ischolten.is@gmail.com>
Co-authored-by: Delmer Reed <delmer814+1@gmail.com>
Co-authored-by: Marc Delagrammatikas <delagrammatikas@gmail.com>
Co-authored-by: Mansi Gandhi <mansimg@gmail.com>
Co-authored-by: Kristin Albright <albright.kristin0@gmail.com>
Co-authored-by: Nick Stalter <nickstalter@gmail.com>

laliconfigcat commented on lib/config_cat/user.ex:
```diff
@@ -0,0 +1,35 @@
+  @enforce_keys :identifier
+  defstruct [:identifier, email: nil, country: nil, custom: nil]
+
+  def new(identifier, other_props \\ []) do
```
Please add the email and country next to the identifier with a default "" value.

laliconfigcat replied:
Also if the identifier is :nil please use "" instead of it

igorescobar replied:
Working on this one 👍

randycoulman replied:
Extracted as #11

laliconfigcat commented on lib/config_cat/user.ex:
```diff
@@ -0,0 +1,35 @@
+  defp normalize(attribute) do
+    attribute
+    |> to_string()
+    |> String.downcase()
```
The properties shouldn't be lowercased (downcased :) ), they properties should match as is. So e.g. Identifier != identifier in the other SDKs.

randycoulman replied:
Extracted as #12

laliconfigcat commented on README.md:
```diff
@@ -1,7 +1,25 @@
 # ConfigCat SDK for Elixir
```
We'd like to use the same format for the readme.md as the other SDKs. 
- Getting started
- Sample/Demo apps
- Polling Modes
- Need help?
- Contributing
- About ConfigCat 

I think it will be enough to adjust this before the release

randycoulman replied:
See #4.  I've added your notes to that issue.

laliconfigcat commented on lib/config_cat.ex:
```diff
@@ -0,0 +1,187 @@
+       when code >= 200 and code < 300 do
+    with {:ok, config} = Jason.decode(body),
+         etag <- extract_etag(headers) do
+      {:ok, %{state | config: config, etag: etag, last_update: now()}}
```
And here comes the magic to handle our new datagovernance feature. Please adjust this with the same logic from here:
https://github.com/configcat/python-sdk/blob/master/configcatclient/configfetcher.py#L74

randycoulman replied:
Extracted to #13

laliconfigcat commented on lib/config_cat.ex:
```diff
@@ -0,0 +1,187 @@
+    end
+  end
+
+  defp schedule_initial_fetch?(%{options: options}) do
```
Please extract all of the refresh logic to a config_fetcher.ex

laliconfigcat replied:
Also please extract the caching to a cache.ex and provide ability to override it with the options

randycoulman replied:
Extracted the first comment to #14. The second comment is already covered by #10.

laliconfigcat commented on lib/config_cat/api.ex:
```diff
@@ -0,0 +1,11 @@
+
+  def process_request_headers(headers) do
+    [
+      {"User-Agent", "ConfigCat-Elixir/m-0.0.1"},
```
Depending on the polling mode this should be:
- manual poll: ConfigCat-Elixir/m-
- auto poll: ConfigCat-Elixir/a-
- lazy loading: ConfigCat-Elixir/l-

Also please fetch the version from the mix.exs if possible.

randycoulman replied:
Extracted to #16 and #17.

Good catch on the version! Since this was part of our app, we didn't have a separate mix.exs for it at the time.  And I forgot about going back to update those prefixes after we added more polling modes.

laliconfigcat commented on lib/config_cat/fetch_policy.ex:
```diff
@@ -0,0 +1,51 @@
+defmodule ConfigCat.FetchPolicy do
```
Please create 3 different files for the different polling modes if possible:
auto_polling_cache_policy.ex
lazy_loading_cache_policy.ex
manual_polling_cache_policy.ex

randycoulman replied:
Extracted to #15

laliconfigcat commented on lib/config_cat/fetch_policy.ex:
```diff
@@ -0,0 +1,51 @@
+    :type,
+    cache_expiry_seconds: 0,
+    poll_interval_seconds: 0
+  ]
```
We will need an on_configuration_changed_callback parameter here and we will need to call it if the configuration changed.

randycoulman replied:
#9

laliconfigcat commented on lib/config_cat.ex:
```diff
@@ -0,0 +1,187 @@
+
+  @base_url "https://cdn.configcat.com"
+  @base_path "configuration-files"
+  @config_filename "config_v4.json"
```
We are using now the new config_v5.json format. There are some breaking changes in that (structure modified).

randycoulman replied:
Extracted to #18

laliconfigcat commented on test/rollout_test.exs:
```diff
@@ -0,0 +1,120 @@
+defmodule ConfigCat.RolloutTest do
```
testmatrix_input_semantic_2.csv and testmatrix_variationId.csv tests are missing.
https://github.com/configcat/python-sdk/blob/master/configcatclienttests/testmatrix_input_semantic_2.csv
https://github.com/configcat/python-sdk/blob/master/configcatclienttests/testmatrix_variationId.csv

randycoulman replied:
Extracted to #19.

Are these new with the v5 config format? They didn't exist in the Ruby SDK when we first wrote our client.

laliconfigcat commented on lib/config_cat.ex:
```diff
@@ -0,0 +1,187 @@
+
+  defp default_options, do: [api: ConfigCat.API, fetch_policy: FetchPolicy.auto()]
+
+  def get_value(key, default_value, user_or_options \\ []) do
```
There is an other get function called get_variation_id, which main purpose is to provide a variation_id which can be sent e.g. to an analytics provider. This feature is currently behind a feature flag in ConfigCat, but it will be turned on soon. You will be able to get the variation_id from the dasboard for a rule/percentage item/default value and match it with your analytics tool.

randycoulman replied:
Extracted to #20

igorescobar commented:
FYI: I've created a branch called `v1.0.0` based on the `main` branch. Let's target all upcoming PRs to this one instead of the `main` branch. After we feel it's time we can merge the `v1.0.0` branch to the `main` one. 👍

igorescobar commented:
@randycoulman The fact you are working on a fork means I can't push to the same PR as you and we can't collaborate on the same PR.

So I pushed your changes (along with some of mine) to this branch instead: 
https://github.com/configcat/elixir-sdk/tree/pr/randycoulman/2

igorescobar commented:
Or, if you prefer, you can try to address the changes @laliconfigcat suggested and I can jump in later.

randycoulman commented:
@laliconfigcat I've gone through all of your comments and extracted them to new issues. Now that we're on a 1.0 branch, can we go ahead and merge this as-is and then start addressing the issues in separate PRs? This is already a pretty big PR and I think it'll be easier to review the various changes separately.  Thanks!
